lms/djangoapps/grades/grade_utils.py

```diff
@@ -1,11 +1,143 @@
 """
 This module contains utility functions for grading.
 """
+from __future__ import absolute_import, unicode_literals
+
+import logging
+import time
 from datetime import timedelta
+
+from django.core.exceptions import ObjectDoesNotExist
 from django.utils import timezone
+from django.utils.translation import ugettext as _
+from opaque_keys.edx.keys import UsageKey
+
+from courseware.models import StudentModule
 from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from student.models import CourseEnrollment
+from super_csv import ChecksumMixin, CSVProcessor, DeferrableMixin
+
 from .config.waffle import ENFORCE_FREEZE_GRADE_AFTER_COURSE_END, waffle_flags
 
+log = logging.getLogger(__name__)
+
+
+class ScoreCSVProcessor(ChecksumMixin, DeferrableMixin, CSVProcessor):
+    """
+    CSV Processor for file format defined for Staff Graded Points
+    """
+    columns = ['user_id', 'username', 'full_name', 'email', 'student_uid',
+               'enrolled', 'track', 'block_id', 'title', 'date_last_graded',
+               'who_last_graded', 'csum', 'last_points', 'points']
+    required_columns = ['user_id', 'points', 'csum', 'block_id', 'last_points']
+    checksum_columns = ['user_id', 'block_id', 'last_points']
+    # files larger than 100 rows will be processed asynchronously
+    size_to_defer = 100
+    max_file_size = 4 * 1024 * 1024
+    handle_undo = False
+
+    def __init__(self, **kwargs):
+        self.now = time.time()
+        self.max_points = 1
+        self.display_name = ''
+        super(ScoreCSVProcessor, self).__init__(**kwargs)
+        self.users_seen = {}
+
+    def get_unique_path(self):
+        return 'csv/state/{}/{}'.format(self.block_id, self.now)
+
+    def validate_row(self, row):
+        valid = super(ScoreCSVProcessor, self).validate_row(row)
+        if valid:
+            valid = row['block_id'] == self.block_id
+            if valid:
+                points = row['points']
+                if points:
+                    try:
+                        valid = float(row['points']) <= self.max_points
+                        if not valid:
+                            self.add_error(_('Points must not be greater than {}.').format(self.max_points))
+                    except ValueError:
+                        self.add_error(_('Points must be numbers.'))
+                        valid = False
+                else:
+                    valid = True
+            else:
+                self.add_error(_('The CSV does not match this problem. Check that you uploaded the right CSV.'))
+        return valid
+
+    def preprocess_row(self, row):
+        if row['points'] and row['user_id'] not in self.users_seen:
+            to_save = {
+                'user_id': row['user_id'],
+                'block_id': self.block_id,
+                'new_points': float(row['points']),
+                'max_points': self.max_points
+            }
+            self.users_seen[row['user_id']] = 1
```

Comment:
It was a set, but I wanted to json serialize the state of the object, and sets aren't serializable. Rather than have special case code to convert sets to lists, I cheated.